game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleSteps.java

```diff
@@ -0,0 +1,242 @@
+package games.strategy.triplea.delegate.battle.steps;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.attachments.TechAttachment;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleStepStrings;
+import games.strategy.triplea.delegate.battle.MustFightBattle.ReturnFire;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import lombok.Builder;
+import lombok.NonNull;
+import org.triplea.java.collections.CollectionUtils;
+
+/** Get the steps that will occurr in the battle */
+@Builder
+public class BattleSteps implements BattleStepStrings {
+
+  final @NonNull Boolean canFireOffensiveAa;
```

Comment:
Have you consider decomposing this class further?  It would likely simplify the testing substantially as well as this class itself.

To illustrate, the main 'get' method of this class would look something like this:
```
final List<String> steps = new ArrayList<>();
steps.addAll(aaBattleStepBuilder.apply(battleStepParameters);
steps.addAll(firstRunBattleStepBuilder.apply(battleStepParameters);
steps.addAll(submergeSubsBattleStepBuilder.apply(battleStepParameters);
:
:
```

Implied in the above, the class properties here would be extracted to a value object that would hold all of the parameters.

To enable test you would want to be able to inject implementations. That could be done by making the class have the following structure:


```
@Builder
public class BattleSteps {
  @Nonnull private final Function<BattleStepParameters, List<String>> aaBattleStepBuilder;
  :
  :

  // This method call it from test code but do nothing with it, just verify you can invoke this and 
  // and have no errors, that everything can be constructed, that will be an integration test.
  static newBattleStepBuilder() {
     return builder()
           .aaBattleStepBuilder(new AaBattleStepBuilder())
           :
           .build();
  }

} 
```

In the above, you can use test code to call `builder()` and inject mock implementations. Then when testing `get` all you need to do is to verify that you invoked each battle step builder, have each one return a stubbed value and then be sure that all of the stubbed values are present on the returned list.

EG:
```

@Mock
private Function<BattleStepParameters, List<String>> aaBatttleStepBuilder;

private BattleSteps battleSteps;

@Mock
private BattleStepParameters battleStepParameters;

@BeforeEach
void setup() {
  battleSteps = BattleSteps.builder()
     .aaBattleStepBuilder(aaBattleStepBuilder)
    :
    :
    .build();
}

@Test
void get() {
when(aaBattleStepBuilder.apply(battleStepParameters)).thenReturn("aaBattleStep"));
:
:

List<String> battleSteps = battleSteps.get(battleStepParameters);

// using explicit static imports as an example, favor statically importing the matchers below
assertThat(battleSteps, IsCollectionWithSize.hasSize( x ));
assertThat(battleSteps, IsIterativeWithItems.hasItems(
   "aaBattleSteps",
   :
   :
   ));
}
```

Finally from there, you can implement each set of battle steps one-by-one and test them each individually.

WDYT?

Reply:
Oh, I absolutely plan to reconstruct this class.  This PR is just the first step.  I wanted a small PR where everything was unittested.  Then I can create further PRs where I start to reconstruct it.

Have you seen my ideas on the forum - https://forums.triplea-game.org/topic/2011/refactoring-mustfightbattle?  I'd like to combine this code with the step executable code.  Each step would be its own separate class, like you mentioned.